### PR TITLE
Fix TraceEnum_ELBO jit tests, etc.

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #1418

- wraps `packed.pack()` in `ignore_jit_warnings`
- avoids `einsum` when not necessary, using `tensordot` instead
- fixes our `einsum` patch to allow multiple `*operands` (high priority fix!)
- tweaks `pyro.ops.einsum.paths` objective function to not break when `opt_einsum` releases their next release

Note: although the jit tests are still marked xfail, they should start to pass as soon as we switch to the upcoming `opt_einsum` 2.3 release.

### CODE REVIEW

File: pyro/ops/einsum/paths.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def _get_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2)
     two = k1 & k2
     one = either - two
     k12 = (either & output) | (two & dim_ref_counts[3]) | (one & dim_ref_counts[2])
-    cost = 2 * _footprint(k12, sizes) - footprints[k1] - footprints[k2]
```
Comment (fritzo) [inline pyro/ops/einsum/paths.py]: This brings makes our objective function match the upstream `eager` objective exactly. We will be deleting this entire file before 0.3 release.

Comment (neerajprad) [reply pyro/ops/einsum/paths.py]: Is the `opt_einsum` scheduled to release soon?

Comment (fritzo) [reply pyro/ops/einsum/paths.py]: Yes, @dgasmith [offered](https://github.com/dgasmith/opt_einsum/pull/74#issuecomment-441390408) to cut a minor release so we could get recent `opt_einsum` improvements into Pyro 0.3.

**Comment by eb8680** [comment]
Looks like some JIT-related test failures in `Dice`.  I wonder why didn't those appear elsewhere already?  Is it from a change in the nightly build?

**Comment by fritzo** [comment]
Hmm weird, the failures only appear in `opt_einsum` release; I've been running `opt_einsum` master locally and tests all pass.

@dgasmith any chance you could cut a minor `opt_einsum` release soon?

**Comment by dgasmith** [comment]
Writing the release notes now. Should cut tomorrow.

**Comment by dgasmith** [comment]
`opt_einsum` 2.3.1 is now on Conda-forge and PyPI.

**Pull request merged**